pulsar-client-cpp/lib/ProducerImpl.cc

```diff
@@ -268,6 +264,14 @@ void ProducerImpl::sendAsync(const Message& msg, SendCallback callback) {
         // If batching is enabled we compress all the payloads together before sending the batch
         payload = CompressionCodecProvider::getCodec(conf_.getCompressionType()).encode(payload);
     }
+    uint32_t compressedSize = payload.readableBytes();
+    if (compressedSize > Commands::MaxMessageSize) {
+        if (conf_.getBlockIfQueueFull()) {
```

Comment:
I guess this should address msb comments - we should be good as long as we reserve the spot before the lock.
https://gist.github.com/jai1/9c013fa1ab174d754108e58ce4cb0e7d